src/async_fuse/memfs/s3_metadata.rs

```diff
@@ -56,11 +60,13 @@ const TXN_RETRY_LIMIT: u32 = 10;
 /// File system in-memory meta-data
 #[derive(Debug)]
 #[allow(dead_code)]
-pub struct S3MetaData<S: S3BackEnd + Send + Sync + 'static> {
+pub struct S3MetaData<S: S3BackEnd + Send + Sync + 'static, St: Storage + Send + Sync + 'static> {
     /// S3 backend
     pub(crate) s3_backend: Arc<S>,
     /// Global data cache
     pub(crate) data_cache: Arc<GlobalCache>,
+    /// Storage manager
+    pub(crate) storage: Arc<StorageManager<St>>,
```

Comment:
I don't think generics are necessary here. Generics should be more about having code run with different types at runtime, like in the case of `max` and `min` func . But in reality, after our configuration, `MetaData` is **no longer generic**. I believe it would be more appropriate to use associated types, or even to hard code the implementation.

In the final version, both Metadata and Node should be non-generic.

Reply:
Perhaps we can refactor all of them in the future, but generic (`St` here) is neccessary for `StorageManager`. If we introduce more storage layers in the future, we may have to define a very complex type named `MemoryCache<LruPolicy<BlockCoordinate>, DiskCache<FifoPolicy<BlockCoordinate>, Backend>>>`, without generic.

Generics make the codes more flexible, and make features such as static dispatch possible (so it can make codes run with different types at **compile time**), and profits from the advanced type system of Rust, it doesn't introduce extra mental burden (unlike C++). In fact, it's very tipical and appropriate to use generic in the case here: the same codes (in `StorageManager` and `S3Metadata`) work with different types (`Storage` implementors). This makes the `Storage`s like some kind of "plugins" (in compile time). 

So I think we should make full use of such a core mechanism, because flexible codes can respond to changing requirements.

Reply:
I'm not saying generics are bad. My point might be about using associated types instead of generics. For example, `Iterator` uses associated types, not generics。

Reply:
The generic parameter of `S3Metadata` is removed. After `S3Backend`'s removal (may be in the next PR), `S3Metadata` and `S3Node` will be with no generic parameters.